File: JavaSort/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    public static void main(String[] args) {

        int[] arr = {2, 6, 29, 17, 3, 91, 70, 61, 20};
        mergeSort(arr, 0, arr.length - 1);
        System.out.println("The Sorted Array is: " + Arrays.toString(arr));
    }
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            // Find the middle point to divide the array into two halves
            int middle = left + (right - left) / 2;

            // Recursively sort the left and right halves
            mergeSort(arr, left, middle);
            mergeSort(arr, middle + 1, right);

            // Merge the sorted halves
            merge(arr, left, middle, right);
        }
    }

    public static void merge(int[] arr, int left, int mid, int right){
        int n1 = mid - left +1;
        int n2 = right -mid;

        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];

        for (int i = 0; i < n1; i++) {
            leftArr[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            rightArr[j] = arr[mid + 1 + j];
        }
        // Merge the two subarrays back into the original array
        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            if (leftArr[i] <= rightArr[j]) {
                arr[k] = leftArr[i];
                i++;
            } else {
                arr[k] = rightArr[j];
                j++;
            }
            k++;
        }

        // Copy any remaining elements from the leftArray, if any
        while (i < n1) {
            arr[k] = leftArr[i];
            i++;
            k++;
        }

        // Copy any remaining elements from the rightArray, if any
        while (j < n2) {
            arr[k] = rightArr[j];
            j++;
            k++;
        }
    }
}

File: PythonSort/mergeSort.py
def merge(left, right):
    merged = []
    left_index, right_index = 0, 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index]< right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1
    merged.extend(left[left_index:])
    merged.extend(right[right_index:])
    
    return merged


def merge_sort(arr, left_index, right_index):
        if left_index < right_index:
            mid = (left_index + right_index)//2
            left = merge_sort(arr, left_index, mid)
            right= merge_sort(arr, mid+1, right_index)
            return merge(left, right)
        return arr[left_index:right_index + 1]

arr = [64,81,97,18,10,891,72,110,69,21,7,-56,32,-7]

sorted_arr = merge_sort(arr, 0, 13)

print("The sorted array is: ", sorted_arr)